File: 卡码/代码随想录/1/08-贪心算法/Python/03-困难/LeetCode_0968.py
#968. 监控二叉树
from typing import *
from ..TreeNode import *

'''
0:无覆盖
1:有摄像头
2:有覆盖
'''
class Solution:
  def __init__(self):
    self.res:int=0
  def traversal(self,root:Optional[TreeNode])->int:
    if root==None:
      return 2
    left:int=self.traversal(root.left)
    right:int=self.traversal(root.right)
    if left==2 and right==2:
      return 0
    elif left==0 or right==0:
      self.res+=1
      return 1
    elif left==1 or right==1:
      return 2
  def minCameraCover(self, root: Optional[TreeNode]) -> int:
    if self.traversal(root)==0:
      self.res+=1
    return self.res

File: 卡码/代码随想录/C++/贪心/03-困难/LeetCode_0968.cpp
//968. 监控二叉树
#include <iostream>
#include "./TreeNode.h"
using namespace std;

/*
 * 0:无覆盖
 * 1:有摄像头
 * 2:有覆盖
 * */
class Solution {
	int res=0;
public:
	int traversal(TreeNode* root){
		if(!root) return 2;
		int left=traversal(root->left);
		int right=traversal(root->right);
		if(left==2&&right==2) return 0;
		else if(left==0||right==0){
			res++;
			return 1;
		}
		else if(left==1||right==1) return 2;
		return -1;
	}
	int minCameraCover(TreeNode* root) {
		if(traversal(root)==0) res++;
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}
